### PR TITLE
add: User identity negotiation

### PR DESCRIPTION
**Pull request by nicholasrussell**
Hello,

Looking for your feedback in adding some of the [user identity negotiation](https://dicom.nema.org/medical/dicom/current/output/chtml/part07/sect_D.3.3.7.html) options.

In order to keep this minimal, I added the entrypoint into retrieving this data into the existing AccessControl trait. The downside being that it breaks the interface.

Thanks

### CODE REVIEW

**Review by slimjim49j — APPROVED** [review]

**Comment by Enet4** [comment]
Hello! Thank you for the pull request! It is great to see an initiative towards the integration on authorization mechanisms in DICOM-rs.

File: ul/src/pdu/mod.rs [hunk 1]
```diff
@@ -330,6 +330,79 @@ pub enum UserVariableItem {
...
+impl UserIdentityType {
+    fn from(user_identity_type: u8) -> Self {
+        match user_identity_type {
+            1 => Self::Username,
+            2 => Self::UsernamePassword,
+            3 => Self::KerberosServiceTicket,
+            4 => Self::SAMLAssertion,
+            5 => Self::JWT,
+            _ => Self::Unsupported,
+        }
```
Comment (Enet4) [inline ul/src/pdu/mod.rs:393]:
```suggestion
    fn from(user_identity_type: u8) -> Option<Self> {
        match user_identity_type {
            1 => Some(Self::Username),
            2 => Some(Self::UsernamePassword),
            3 => Some(Self::KerberosServiceTicket),
            4 => Some(Self::SamlAssertion),
            5 => Some(Self::Jwt),
            _ => None,
        }
```

File: ul/src/pdu/mod.rs [hunk 2]
```diff
@@ -330,6 +330,79 @@ pub enum UserVariableItem {
...
+    }
+
+    fn to_u8(&self) -> u8 {
+        match self {
+            Self::Username => 1,
+            Self::UsernamePassword => 2,
+            Self::KerberosServiceTicket => 3,
+            Self::SAMLAssertion => 4,
+            Self::JWT => 5,
+            Self::Unsupported => 0,
```
Comment (Enet4) [inline ul/src/pdu/mod.rs:403]:
```suggestion
            Self::SamlAssertion => 4,
            Self::Jwt => 5,
```

File: ul/src/pdu/reader.rs [hunk 3]
```diff
@@ -945,6 +945,52 @@ where
...
+
+                        // n+3-m - Secondary Field
+                        let secondary_field = read_n(&mut cursor, secondary_field_length as usize)
+                            .context(ReadPduFieldSnafu {
+                                field: "User-Identity-secondary-field",
+                            })?;
+
+                        user_variables.push(UserVariableItem::UserIdentityItem(UserIdentity::new(
+                            positive_response_requested == 1,
+                            UserIdentityType::from(user_identity_type),
```
Comment (Enet4) [inline ul/src/pdu/reader.rs:989]: Here we would either change `UserIdentity` to cover the possibility of an unknown user identity type code, or skip the variable with a warning.

**Review by Enet4 — CHANGES_REQUESTED** [review]
Hello @nicholasrussel! I am adding this contribution as a candidate for DICOM-rs 0.7.0, but we may need a reiteration of the code first. Please see the suggestions inline, and let me know your availability to work on this.

If you also have any ideas for testing the changes in a real scenario, I would be grateful to know. The PACS archives I usually work with do not feature user identity negotiation.

File: ul/src/association/client.rs [hunk 4]
```diff
@@ -270,6 +285,76 @@ impl<'a> ClientAssociationOptions<'a> {
         self
     }
 
+    /// Sets the user identity username
+    pub fn username<T>(mut self, username: T) -> Self
+    where
+        T: Into<Cow<'a, str>>,
+    {
+        let username = username.into();
+        if username.is_empty() {
```
[the remark this answers is not in the document]
Comment (Enet4) [reply ul/src/association/client.rs:294]: Never mind, I read this incorrectly.

**Review by Enet4 — APPROVED** [review]
Despite there being gaps in the ecosystem for full authentication support, there is interest in expanding the `ul` crate with the knowledge of user identity parts, plus the extra of making these available in the `dicom-storescu` tool. This appears to be in order, so I will merge it onto the 0.7 milestone. Much appreciated! :+1:

**Pull request merged**